### ISSUE TITLE
[FEAT]: Preserve JsonProperty in extended class

### ISSUE THREAD

**Issue by devhl-labs** [GillianPerard/typescript-json-serializer#227]
### Description

When extending a class, the JsonProperties from the base class are dropped. Is it possible to keep those intact so serialization continues to work?

**Comment by GillianPerard** [comment]
Hi, could you provide an example what you suggest is the normal behavior...

**Comment by devhl-labs** [comment]
I am seeing something different when I write the example. This is confusing me because I thought a property without JsonProperty gets ignored. Bar b property does not have JsonProperty but it is getting serialized in this example.

```typescript
@JsonObject()
export class Foo {
    @JsonProperty()
    readonly a: string

    constructor(a: string) {
        this.a = "a"
    }
}

export class Bar extends Foo {
    readonly b: string

    constructor(b: string) {
        super("a")
        this.b = b
    }
}

describe('Extending a class', () => {
    it('serializes the base', () => {

        const foo = new Foo("a")

        const bar = new Bar("b")

        const serializer = new JsonSerializer()

        const expected = serializer.serialize(foo)

        const actual = serializer.serialize(bar)

        assert.strictEqual(actual, expected) // expected Bar{ a: 'a', b: 'b' } to equal Foo{ a: 'a' }
    })
})
```

**Comment by devhl-labs** [comment]
It looks like things are different when there is a base class. The serializer returns an empty object. I think the answer is to remove JsonObject from the all but the base class. But then it includes properties that do not have JsonProperty, which I am not sure why.

```typescript
@JsonObject()
class Base {
    @JsonProperty()
    readonly name: string

    constructor(name: string) {
        this.name = name
    }
}

@JsonObject()
export class Foo extends Base {
    @JsonProperty()
    readonly a: string

    constructor(a: string) {
        super("base")
        this.a = "a"
    }
}

export class Bar extends Foo {
    readonly b: string

    constructor(b: string) {
        super("a")
        this.b = b
    }
}

describe('Extending a class', () => {
    it('serializes the base', () => {

        const foo = new Foo("a")

        const bar = new Bar("b")

        const serializer = new JsonSerializer()

        const expected = serializer.serialize(foo)

        const actual = serializer.serialize(bar)

        assert.strictEqual(actual, expected) // expected {} to equal {}
    })
})
```

**Comment by GillianPerard** [comment]
I don't reproduce the same result you obtain in your examples...
But I think you miss the `JsonObject()` on `Bar` class.

Which platform (Windows/Linux/MacOs) and techno (Node, React, Vue, Angular, and so on) are you using?

**Comment by devhl-labs** [comment]
Using Ubuntu via WSL with node v18.20.3. When you tested did you have a base class? I get an empty object when serializing Foo and Bar.

```ts
@JsonObject()
class Base {
    @JsonProperty()
    readonly name: string

    constructor(name: string) {
        this.name = name
    }
}

@JsonObject()
export class Foo extends Base {
    @JsonProperty()
    readonly a: string

    constructor(a: string) {
        super("base")
        this.a = "a"
    }
}

export class Bar extends Foo {
    readonly b: string

    constructor(b: string) {
        super("a")
        this.b = b
    }
}
```

**Comment by devhl-labs** [comment]
I tried making a sample repo, but I also did not see the same thing. No idea what I did wrong, but I think I found a work around anyway. FYI, the initial request was to be able to serialize an instance of Bar and IGNORE any properties that the end user added. So your attributes would not have been on Bar, only Base and Foo. The intent was that serializing Bar would be the same as serializing Foo. I can explain the reason if your intrested.
